sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/QueryPlan.scala

```diff
@@ -17,16 +17,34 @@
 
 package org.apache.spark.sql.catalyst.plans
 
-import org.apache.spark.sql.catalyst.expressions.{Attribute, AttributeSet, Expression, VirtualColumn}
+import org.apache.spark.sql.catalyst.expressions._
 import org.apache.spark.sql.catalyst.trees.TreeNode
 import org.apache.spark.sql.types.{DataType, StructType}
 
-abstract class QueryPlan[PlanType <: TreeNode[PlanType]] extends TreeNode[PlanType] {
+abstract class QueryPlan[PlanType <: TreeNode[PlanType]]
+  extends TreeNode[PlanType] with PredicateHelper {
   self: PlanType =>
 
   def output: Seq[Attribute]
 
   /**
+   * Extracts the relevant per-row constraints from a given child while removing those that
+   * don't apply anymore.
+   */
+  protected def getRelevantConstraints(child: QueryPlan[PlanType]): Set[Expression] = {
+    child.constraints.filter(_.references.subsetOf(outputSet))
+  }
+
+  /**
+   * A sequence of expressions that describes the data property of the output rows of this
+   * operator. For example, if the output of this operator is column `a`, an example `constraints`
+   * can be `Set(a > 10, a < 20)`.
+   */
+  lazy val constraints: Set[Expression] = validConstraints
```

Comment:
I would consider making this `getRelevantConstraints(validConstraints)` so that each implementor of `validContraints` does't have to remember to do the filter / canonicalization.  They can just focus on augmenting or passing through constraints from children based on the operators semantics.